Pia 1642

zladzeyka opened the pull request:
### Description
CameraPreviewViewController: added loading indicator while camera is setting up (PIA-1642)
Camera: configureSession with input and output setup and also qrOutput are run a dedicated serial dispatch queue (sessionQueue) for not blocking the main queue (PIA-1642)
Camera: added updated method AVCapturePhotoCaptureDelegate @available(iOS 11.0, *) (PIA-1642)

### How to test


### Merging
Automatic

### Issues
(PIA-1642)

zladzeyka committed:
Camera: added updated method AVCapturePhotoCaptureDelegate  @available(iOS 11.0, *) (PIA-1642)

zladzeyka committed:
Camera: configureSession with input and output setup and also qrOutput are run a dedicated serial dispatch queue (sessionQueue) for not blocking the main queue (PIA-1642)

zladzeyka committed:
GiniConfiguration: added camera loading indicator color (PIA-1642)

zladzeyka committed:
CameraPreviewViewController: added loading indicator while camera is setting up (PIA-1642)

zladzeyka committed:
CameraPreviewView: updated initialiser for layerClass (PIA-1642)

zladzeyka committed:
Camera: updated initialiser for serial camera sessionQueue (PIA-1642)

zladzeyka committed:
Renamed methods for loading/stopLoading indicator (PIA-1642)

a-szotyori commented on GiniCapture/Classes/Core/Screens/Camera/Camera.swift:
```diff
@@ -300,4 +311,16 @@ extension Camera: AVCapturePhotoCaptureDelegate {
+            didCaptureImageHandler?(nil, .captureFailed)
+            return
+        } else {
+            let photoData = photo.fileDataRepresentation()
```
The documentation for `fileDataRepresentation()` says that the image data format is set using the `AVCapturePhotoSettings` passed to `AVCapturePhotoOutput.capturePhoto(with:delegate)`. Currently we don't specify the format when creating the `AVCapturePhotoSettings` so we depend on Apple which format is used. I checked and currently it's jpeg.

We should explicitly set the format when creating the `AVCapturePhotoSettings` to make sure it's jpeg:
```
let captureSettings: AVCapturePhotoSettings
if #available(iOS 11.0, *) {
    captureSettings = AVCapturePhotoSettings(rawPixelFormatType: 0,
                                             rawFileType: nil,
                                             processedFormat: nil,
                                             processedFileType: AVFileType.jpg)
} else {
    captureSettings = AVCapturePhotoSettings()
}```

zladzeyka replied:
good point, thanks

a-szotyori commented on GiniCapture/Classes/Core/Screens/Camera/CameraViewController.swift:
```diff
@@ -446,6 +446,7 @@ extension CameraViewController: CameraPreviewViewControllerDelegate {
         cameraButtonsViewController.isFlashSupported = camera.isFlashSupported
         cameraButtonsViewController.view.setNeedsLayout()
         cameraButtonsViewController.view.layoutIfNeeded()
+        viewController.stopLoadingIndicator()
```
Stopping the loading indicator could be done by the `CameraPreviewViewController` itself before or after calling `self.delegate?.cameraDidSetUp(self, camera: self.camera)` on line 169. This would encapsulate the loading indicator starting/stopping logic in `CameraPreviewViewController`.

zladzeyka replied:
🙈

a-szotyori reviewed (commented):
Looks good. I have only two observations.

zladzeyka committed:
Review notes: set JPEG format in  AVCapturePhotoSettings (PIA-1642)

zladzeyka committed:
Review notes: CameraPreviewController fixed logic encapsulation regarding to loading indicator (PIA-1642)

zladzeyka committed:
Camera: qrCode output setup moved in camera (PIA-1642)

zladzeyka committed:
Camera: added check for optional spinner PIA-1642)

zladzeyka committed:
CameraPreviewViewController: fixed test setupQRCodeOutput (PIA-1642)

a-szotyori commented on GiniCapture/Classes/Core/Screens/Camera/CameraPreviewViewController.swift:
```diff
@@ -198,7 +197,9 @@ final class CameraPreviewViewController: UIViewController {
     func stopLoadingIndicator(){
         DispatchQueue.main.async {
-            self.spinner.stopAnimating()
+            if (self.spinner != nil){
```
This made me think 🤔. Wouldn't it be better to create the spinner in a lazy computed property like the `previewView` and add it in `loadView()`? Then it can never be `nil` and you can simply add it as a subview without needing to dispatch to the main queue.

zladzeyka replied:
Done . Good point👍🏻

a-szotyori reviewed (commented):
One small question.

zladzeyka committed:
Review notes:
CameraPreviewViewController: added spinner as a lazy property, remove dispatch blocks (PIA-1642)
